### PR TITLE
Persistent storage proposal

### PR DESCRIPTION
**Pull request by markturansky**
Breaking up #4055 into smaller PRs for focused discussion.

Subsequent PRs will be for the API and implementation (volume manager and plugin).

@thockin @bgrant0607 @smarterclayton @pmorie

### CODE REVIEW

**Comment by justinsb** [comment]
This is very clear - thank you for taking the time to split it out from #4055.

I'm not sure it makes sense to punt on the dynamic creation issue - that seems like it will be a very common use case (AWS, GCE, OpenStack, SANs).  What is the plan here?  I can see two options:

1) The cloud provider dynamically creates a volume when one is claimed.  (And if so, how does the client know that a volume will magically appear on-demand so it doesn't block the request because it sees there are no volumes available?) 
2) the caller creates a volume (with an API call) before creating their claim.

I think a worked example of launching a pod with a volume in this scenario would clear up my confusion here!

**Comment by markturansky** [comment]
We're not punting on dynamic provisioning, just putting the API in place first for Persistent Volumes.  The horse goes before the cart.

Our experience with Open Shift is that creating and formatting a volume on the fly is slow.  Better to have a pool of readily available volumes to mount, at least at the present time.  As that process gets better, though, the time between "create" and "consume" will decrease, possibly enough to allow on the fly creation.

The dynamic provisioning of volumes is not yet specified.

**Comment by thockin** [comment]
I expect dynamic provisioning will come soon after this is all integrated.  I think THIS is a 1.0 blocker, but dynamic is probably not.

**Comment by markturansky** [comment]
@thockin agreed because an admin can script the creation of the volumes.  dynamic provisioning eases their pain.

File: docs/design/persistent-storage.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+Cluster adminstrators use the API to manage *PersistentVolumes*.  The singleton PersistentVolumeManager watches the Kubernetes API for new volumes and adds them to its internal cache of volumes in the system. All persistent volumes are managed and made available by the volume manager.  The manager also watches for new claims for storage and binds them to an available volume by matching the volume's characteristics (AccessModes and storage size) to the user's request.
+
+Many means of dynamic provisioning will be eventually be implemented for various storage types. 
+
+
+##### PersistentVolume API
+
+| Action | HTTP Verb | Path | Description |
+| ---- | ---- | ---- | ---- |
+| CREATE | POST | /api/{version}/persistentvolumes/ | Create instance of PersistentVolume in system namespace  |
```
Comment (derekwaynecarr) [inline docs/design/persistent-storage.md]: There is no system namespace.  If I understand correctly, a PersistentVolume is not in any namespace - i.e it follows the same convention as a Node today.  If that is the case we need to remove mention of a system namespace here or define what the system namespace means before this section.

**Comment by derekwaynecarr** [comment]
Can you clarify if a PersistentVolume drives Events in the system?  If so, can we assert that any events it spawns will be treated consistent with Node objects.  Meaning they will go in the cluster-id namespace?

**Comment by derekwaynecarr** [comment]
And because I was bit by this before, why did we merge a PR that was WIP?

**Comment by thockin** [comment]
I forgot to edit the subject, but you have good feedback.  Mark can get
that feedback wrapped up as a new PR :)

**Pull request merged**